pyqtorch/utils.py

```diff
@@ -116,3 +116,20 @@ def _rand(n_qubits: int) -> torch.Tensor:
 
 def param_dict(keys: Sequence[str], values: Sequence[torch.Tensor]) -> dict[str, torch.Tensor]:
     return {key: val for key, val in zip(keys, values)}
+
+
+def density_mat(
+    vector_state: torch.Tensor, dtype: torch.dtype = DEFAULT_MATRIX_DTYPE
+) -> torch.Tensor:
+    """
+    Transforms a pure state :math:`|\\psi\\rangle` into its corresponding density matrix
+    :math:`\\rho = |\\psi\\rangle \\langle\\psi|`.
+    """
+    n_qubits = len(vector_state.size()) - 1
+    batch_size = vector_state.size(-1)
+    density_mat = torch.zeros(2**n_qubits, 2**n_qubits, batch_size, dtype=dtype)
+    for i in range(batch_size):
+        v_state_ket = vector_state.reshape((2**n_qubits, batch_size))[:, i]
+        density = torch.outer(v_state_ket, v_state_ket.conj())
+        density_mat[:, :, i] = density
```

Comment:
You can use `torch.stack` here.

Reply:
By using it I get this:

```python
def density_mat(
    vector_state: torch.Tensor, dtype: torch.dtype = DEFAULT_MATRIX_DTYPE
) -> torch.Tensor:
    """
    Transforms a pure state :math:`|\\psi\\rangle` into its corresponding density matrix
    :math:`\\rho = |\\psi\\rangle \\langle\\psi|`.
    """
    n_qubits = len(vector_state.size()) - 1
    batch_size = vector_state.size(-1)
    density_list = []
    for i in range(batch_size):
        v_state_ket = Vector_state.reshape((2**n_qubits, batch_size))[:, i]
        density = torch.outer(v_state_ket, v_state_ket.conj())
        density_list.append(density)

    density_mat = torch.stack(density_list, dim=2)
    return density_mat```

 I'm not sure it's better. What do you think ?

Reply:
Yes. I think it is best to rely on torch functionalities as they are efficient.

Reply:
I wrote a new version of the function using your various comments and using @dominikandreasseitz's comment [here](https://github.com/pasqal-io/pyqtorch/pull/148/files/b747fcb0ec802b126aea07f5266673858d97fbb4#r1535225671 ):

```python
def density_mat(
    vector_state: torch.Tensor, dtype: torch.dtype = DEFAULT_MATRIX_DTYPE
) -> torch.Tensor:
    
    #Permute the vector's batch first:
    batch_dim = vector_state.dim() - 1  
    batch_first_perm = [batch_dim] + list(range(batch_dim))
    vector_state_perm = torch.permute(vector_state, batch_first_perm)

    #reshape using flatten:
    vector_state_perm_flat = vector_state_perm.flatten()

    # Split for every batch:
    vector_state_perm_flat_split = torch.split(vector_state_perm_flat,split_size_or_sections=2**n_qubits)

    #Compute the permute projector and stack it: 
    proj_list = []
    for ket in vector_state_perm_flat_split:
        proj_ket = torch.outer(ket,ket.conj())
        proj_list.append(proj_ket)
    projector_perm = torch.stack(proj_list)

    #Permute the matrix's batch last:
    batch_first_perm_mat = (2, 0, 1)
    undo_perm = tuple(torch.argsort(torch.tensor(batch_first_perm_mat)))
    return torch.permute(projector_perm, undo_perm)
```

What do you think ?

Reply:
Looking good to me. I would somehow rename variables but we can discuss once you've implemented it.